allennlp/nn/decoding/maximum_marginal_likelihood.py

```diff
@@ -33,43 +34,50 @@ def decode(self,
                decode_step: DecoderStep,
                targets: torch.Tensor = None,
                target_mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
-        if targets.dim() != 2:
-            raise NotImplementedError("This implementation cannot yet handle batched inputs")
         allowed_transitions = self._create_allowed_transitions(targets, target_mask)
         finished_states = []
         states = [initial_state]
         step_num = 0
         while states:
             step_num += 1
             next_states = []
-            for state in states:
-                allowed_actions = allowed_transitions[tuple(state.action_history)]
-                generator = decode_step.take_step(state, allowed_actions)
-                while True:
-                    try:
-                        next_state = next(generator)
-                    except StopIteration:
-                        break
-                    if next_state.is_finished():
-                        finished_states.append(next_state)
-                    else:
-                        next_states.append(next_state)
+            # We group together all current states to get more efficient (batched) computation.
+            grouped_state = states[0].combine_states(states)
+            allowed_actions = self._get_allowed_actions(grouped_state, allowed_transitions)
+            generator = decode_step.take_step(grouped_state, allowed_actions)
+            while True:
+                try:
+                    next_state = next(generator)
+                except StopIteration:
+                    break
+                finished, not_finished = next_state.split_finished()
+                if finished is not None:
+                    finished_states.append(finished)
+                if not_finished is not None:
+                    next_states.append(not_finished)
             states = next_states
 
-        scores = torch.cat([state.score for state in finished_states])
-        return {'loss': -util.logsumexp(scores)}
+        # This is a dictionary of lists - for each batch instance, we want the score of all
+        # finished states.  So this has shape (batch_size, num_target_action_sequences), though
+        # it's not actually a tensor, because different batch instance might have different numbers
+        # of finished states.
+        batch_scores = self._group_scores_by_batch(finished_states)
+        loss = 0
+        for scores in batch_scores.values():  # we don't care about the batch index, just the scores
+            loss += -util.logsumexp(torch.cat(scores))
+        return {'loss': loss / len(batch_scores)}
 
     @staticmethod
     def _create_allowed_transitions(targets: torch.Tensor,
-                                    target_mask: torch.Tensor = None) -> Dict[Tuple[int, ...], Set[int]]:
```

Comment:
These arguments don't seem to be optional at the moment. Maybe just remove these defaults for now?

Reply:
Done.